tango/integrations/wandb/workspace.py

```diff
@@ -0,0 +1,429 @@
+import logging
+import tempfile
+from datetime import datetime
+from pathlib import Path
+from typing import Any, Dict, Iterable, Iterator, Optional, TypeVar, Union
+from urllib.parse import ParseResult
+
+import pytz
+import wandb
+
+from tango.common.file_lock import FileLock
+from tango.common.util import exception_to_string, tango_cache_dir, utc_now_datetime
+from tango.step import Step
+from tango.step_cache import StepCache
+from tango.step_info import StepInfo, StepState
+from tango.workspace import Run, Workspace
+
+from .step_cache import WandbStepCache
+from .util import ArtifactKind, RunKind, check_environment
+
+T = TypeVar("T")
+
+logger = logging.getLogger(__name__)
+
+
+@Workspace.register("wandb")
+class WandbWorkspace(Workspace):
+    """
+    This is a :class:`~tango.workspace.Workspace` that tracks Tango runs in a W&B project.
+    It also stores step results as W&B Artifacts via :class:`WandbStepCache`.
+
+    Each Tango run with this workspace will generate multiple runs in your W&B project.
+    There will always be a W&B run corresponding to each Tango run with the same name,
+    which will contain some metadata about the Tango run. Then there will be one W&B run
+    for each cacheable step that runs with a name corresponding to the name of the step.
+    So if your Tango run includes 3 cacheable steps, that will result in a total of 4 new runs in W&B.
+
+    :param project: The W&B project to use for the workspace.
+    :param entity: The W&B entity (user or organization account) to use for the workspace.
+
+    .. tip::
+        Registered as a :class:`~tango.workspace.Workspace` under the name "wandb".
+    """
+
+    def __init__(self, project: str, entity: Optional[str] = None):
+        check_environment()
+        super().__init__()
+        self.project = project
+        self._entity = entity
+        self.cache = WandbStepCache(project=self.project, entity=self.entity)
+        self.steps_dir = tango_cache_dir() / "wandb_workspace"
+        self.locks: Dict[Step, FileLock] = {}
+        self._running_step_info: Dict[str, StepInfo] = {}
+
+    def __getstate__(self):
+        """
+        We override `__getstate__()` to customize how instances of this class are pickled
+        since we don't want to persist certain attributes.
+        """
+        out = super().__getstate__()
+        out["locks"] = {}
+        return out
+
+    @property
+    def wandb_client(self) -> wandb.Api:
+        overrides = {"project": self.project}
+        if self._entity is not None:
+            overrides["entity"] = self._entity
+        return wandb.Api(overrides=overrides)
```

Comment:
I think it's okay, but here's a better solution: a39c5ec

Reply:
🤦‍♂️ turns out my original implementation was necessary, caching it via https://github.com/allenai/tango/commit/a39c5ece0f8d1605ff9993cc1da57de23f7be529 actually breaks things since `wandb.Api` does some internal caching of HTTP requests to the server. I finally figured out that this was causing the bug with step state being out-of-date. FYI @AkshitaB 

Reverted in 5894941